lgc/builder/BuilderImpl.cpp

```diff
@@ -688,109 +741,95 @@ Instruction *BuilderImpl::createWaterfallLoop(Instruction *nonUniformInst, Array
   SmallVector<Value *, 2> nonUniformIndices;
   // Maps the nonUniformIndex that is returned by traceNonUniformIndex() to the nonUniformInst.
   DenseMap<Value *, std::pair<Value *, unsigned>> nonUniformIndexImageCallOperand;
-  TraceNonUniformIndex traceNonUniformIndex(nonUniformInst, scalarizeDescriptorLoads, 64);
+  TraceNonUniformIndex traceNonUniformIndex(nonUniformInst, scalarizeDescriptorLoads);
+  DenseMap<unsigned, Value *> operandIdxnonUniformIndex;
 
   for (unsigned operandIdx : operandIdxs) {
     Value *nonUniformImageCallOperand = nonUniformInst->getOperand(operandIdx);
     Value *nonUniformIndex = traceNonUniformIndex.run(nonUniformImageCallOperand);
-    scalarizeDescriptorLoads = traceNonUniformIndex.foundDependencies();
     if (nonUniformIndex) {
       nonUniformIndices.push_back(nonUniformIndex);
 
-      if (scalarizeDescriptorLoads)
+      if (scalarizeDescriptorLoads) {
         nonUniformIndexImageCallOperand[nonUniformIndex] = std::make_pair(nonUniformImageCallOperand, operandIdx);
+        operandIdxnonUniformIndex[operandIdx] = nonUniformIndex;
+      }
     }
   }
 
   if (nonUniformIndices.empty())
     return nonUniformInst;
 
+  // Save Builder's insert point
+  IRBuilder<>::InsertPointGuard guard(*this);
+  // Insert new code just before nonUniformInst.
+  SetInsertPoint(nonUniformInst);
+
   // For any index that is 64 bit, change it back to 32 bit for comparison at the top of the waterfall loop.
-  // At this point the nonUniformVal of nonUniformIndices might change. We also need the original non uniform values for
-  // the scalarization of the descriptor loads.
-  DenseMap<Value *, Value *> newOrigNonUniformVal;
-  for (Value *&nonUniformVal : nonUniformIndices) {
-    if (nonUniformVal->getType()->isIntegerTy(64)) {
-      auto sExt = dyn_cast<SExtInst>(nonUniformVal);
-      Value *origNonUniformVal = nonUniformVal;
-      // 64-bit index may already be formed from extension of 32-bit value.
-      if (sExt && sExt->getOperand(0)->getType()->isIntegerTy(32)) {
-        nonUniformVal = sExt->getOperand(0);
-        newOrigNonUniformVal[nonUniformVal] = origNonUniformVal;
-      } else {
-        nonUniformVal = CreateTrunc(nonUniformVal, getInt32Ty());
-        newOrigNonUniformVal[nonUniformVal] = origNonUniformVal;
-      }
+  DenseMap<Value *, Value *> nonUniformIndex32BitVal;
+  for (auto nonUniformIndex : nonUniformIndices) {
+    // Start the waterfall loop using the waterfall index.
+    if (nonUniformIndex->getType()->isIntegerTy(64)) {
```

Comment:
Can `nonUniformIndex` be a 16-bit value? If so, then it needs to be handled too.

Reply:
I do not know about this. The original code does not take it into consideration. @nhaehnle @perlfu : What do you think about this?